### PR TITLE
Add support for Compose and dead-keys with libxkbcommon

### LINKED ISSUE #4: i3lock don't handle compose and dead-keys
Because of this reason, i3lock can't handle passwords with accents (e.g. á, ë) or other complex glyphs which are composed with 2 or more key presses.

This make the application pretty unusable for that kind of passwords.

### PR DESCRIPTION
**Pull request by danielotero**
This PR should resolve #4

### CODE REVIEW

File: i3lock.c [hunk 1]
```diff
@@ -289,13 +323,36 @@ static void handle_key_press(xcb_key_press_event_t *event) {
...
+        switch (xkb_compose_state_get_status(xkb_compose_state)) {
+        case XKB_COMPOSE_NOTHING:
+            break;
+        case XKB_COMPOSE_COMPOSING:
+            return;
+        case XKB_COMPOSE_COMPOSED:
+            /* xkb_compose_state_get_utf8 doesn't include the terminating byte in the return value
+             * as xkb_keysym_to_utf8 does. Adding one makes the variable n consistent. */
+            n = xkb_compose_state_get_utf8(xkb_compose_state, buffer, sizeof(buffer)) + 1;
+            //ksym = xkb_compose_state_get_one_sym(xkb_compose_state); // Not sure if it's needed
```
Comment (stapelberg) [inline i3lock.c]: Without this line, people cannot compose return/enter/xf86screensaver, right? I think it should be in the code.

File: i3lock.c [hunk 2]
```diff
@@ -138,6 +141,37 @@ static bool load_keymap(void) {
 }
 
 /*
+ * Loads the XKB compose table from the given locale.
+ *
+ */
+static bool load_compose_table(const char *locale) {
+    if (xkb_context == NULL) {
```
Comment (stapelberg) [inline i3lock.c]: You can delete line 148 to 153, xkb_context is guaranteed to be non-NULL at this point (it’s initialized in `load_keymap` and i3lock exits if that fails).

**Comment by stapelberg** [comment]
Very nice!

File: i3lock.c (hunk 2 above)
Comment (danielotero) [inline i3lock.c]: Ok. I just use `load_keymap` as a template, because I never use the libxkbcommon.

File: i3lock.c (hunk 1 above)
Comment (danielotero) [inline i3lock.c]: The function `xkb_keysym_to_utf8` is moved below and is executed only if the current ksym doesn't trigger the compose state machine. In fact, is executed even if the compose combination is cancelled, which is different from what X does. I should fix that too.

Comment (stapelberg) [inline i3lock.c]: > The return/enter/xf86screensaver works for me as expected and I'm not sure that makes sense compose that kind of keys.

Why not? You never know what weird keyboard layouts people have… :)

> And I reallize that the key still gets readed after a failed composing attempt, which is different from what X does. I should fix that too.

I interpret that as you’re not quite done with this pull request yet, then.

Comment (danielotero) [inline i3lock.c]: > Why not? You never know what weird keyboard layouts people have… :)

You are right. I never thought about a composing other than accents.

Now the line is uncommented and the behaviour is like the X server. If no other problem arises, I think it's done.

**Pull request merged**